File: LeetCode/uglyNumber.js
/**
 * @param {number} num
 * @return {boolean}
 */

// Accepted solution
const isUgly = num => {
    for (let divisor of [2, 3, 5]) {
        while (num && num % divisor === 0) {
            num /= divisor;
        }
    }

    return num === 1;
};

// Too slow solution
function getPrimes(n) {
    var listOfNum =range(n),
        i = 2;

    // CHeck only until the square of the prime is less than number
    while (i*i < n && i < n) {
        listOfNum = filterMultiples(listOfNum, i);
        i++;
    }

    return listOfNum;


    function range (num) {
        var res = [];
        for (var i = 2; i <= num; i++) {
            res.push(i);
        }
        return res;
    }

    function filterMultiples (list, x) {
        return list.filter(function (item) {
            // Include numbers smaller than x as they are already prime
            return (item <= x) || (item > x && item % x !== 0);
        });
    }
}

var isUglyTooSlowTLE = function(num) {
    let cache = {};

    var isPrime = function(num) {
        //console.log("Cache from isPrime:", cache);
        if (num in cache) return cache[num];

        if (num <= 2) {
            return cache[num] = true;
        }
        if (num % 2 === 0) {
            return cache[num] = false;
        }

        for (let i = 3; i <= Math.sqrt(num); i+=2) {
            if ( num % i === 0 ) {
                return cache[num] = false;
            }
        }

        return cache[num] = true;
    };

    if (num <= 0) { return false; }
    if (num >= 1 && num <= 6) { return true; }
    if (num === 7) { return false; }

    let primes = getPrimes(num / 2);
    console.log(primes);

    return primes.filter(n => (n !== 2 && n !== 3 && n !== 5) &&  (num/n) === parseInt(num/n) ).length == 0 && !isPrime(num);
};


//////////////
// TESTING
//////////////

//let res = isUgly(22);
//let res = isUgly(905391974);
//let res = isUgly(214797179);
let res = isUgly(1641249143);
//let res = isUgly(28);
console.log(res);

File: LeetCode/uglyNumber.py
# """Write a program to check whether a given number is an ugly number.

# Ugly numbers are positive numbers whose prime factors only include 2, 3, 5. For example, 6, 8 are ugly while 14 is not ugly since it includes another prime factor 7.

# Note that 1 is typically treated as an ugly number."""
# def findPrimes(numToFind):
#     def findPrime(n):
#         for j in range(3, n//2, 2):
#             print("j:", j)
#             if (n / j) % 1 != 0: return False
#         return True

#     primes = []
#     for i in range(3, numToFind // 2, 2):
#         print("i:", i)
#         if findPrime(i): primes.append(i)
#     print(primes)


# def isPrime(n):
#     for i in range(2, n//2 + 1):
#         if (n/i) % 1 == 0: return False
#     return True

# def findPrimes(n):
#     primes = []
#     for i in range(7, n//2):
#         if isPrime(i): primes.append(i)
#     return primes

class Solution(object):
    def isUgly(self, num):
        if num <= 0: return False
        if num == 1: return True
        for divisor in [2, 3, 5]:
            # if (num / i) % 1 == 0: return True
            while num and num % divisor == 0:
                num /= divisor
        return num == 1
        
        # primes = findPrimes(num)
        # for p in primes:
        #     if (num/p) % 1 == 0: return False
        # return True


# findPrimes(1000)
res = Solution().isUgly(2147483649)
print(res)










# def isPrime(n):
#     for i in range(2, n//2 + 1):
#         if (n/i) % 1 == 0: return False
#     return True

# def findPrimes(n):
#     primes = []
#     for i in range(7, n):
#         if isPrime(i): primes.append(i)
#     return primes


#res = isPrime(9)
# res = findPrimes(100)
# print(res)